### PR TITLE
Clear payment account payload info from closed trades.

### LINKED ISSUE #5396: Encrypt or remove saved trader chats and trade data on local Bisq instances
### Description

I have created this issue from the discussion on the Bisq community for forum [Is trader chat saved?](https://bisq.community/t/is-trader-chat-saved/10539/4).

Currently the following information is avlaiable on users local Bisq instances

- Unencrypted trader chats between themselves and peers
- Unencrypted trade data (names, account numbers, trade amounts etc) between themselves and peers

I am unsure what trade information is unencrypted on mediators' or arbitrators' Bisq instances. 

Having trade chats and trade data saved on local Bisq instances is a security concern for both traders and everyone they have traded with.

Having trade chats and trade data saved on mediators' or arbitrators' Bisq instances is a security concern for everyone they have mediated / arbitrated.

Traders with lots of trades, mediators and arbitrators will end up being a centralized source of unencrypted data. This puts users of Bisq at risk.

#### Version

v1.6.2

### Steps to reproduce

open \Bisq\btc_mainnet\db\

- ClosedTrades
- FailedTrades
- MaiboxMessageList

There might be more. I have not checked all the files for unencrypted data.

### Expected behaviour

Chat and trade data to be encrypted.

Not sure if there should be a time limit for how long this data is kept?

### Actual behaviour

Chats and trade data are stored unencrypted.

### Screenshots

Taken from: https://bisq.community/t/is-trader-chat-saved/10539/14

```
SEPA���������:�
>XXXCENSORED_BANK COUNTRY CODE_XXX"�
XXXCENSORED_Full name_XXXeXXXCENSORED_IBAN_XXXXXXCENSORED_bic_XXX*AT*BE*BG*CH*CY*CZ*DE*DK*EE*ES*FI*FR*GB*GR*HR*HU*IE*IS*IT*LI*LT*LU*LV*MC*MT*NL*NO*PL*PT*RO*SE*SI*SKzH
salt@XXXCENSORED_XXX*XXXCENSORED_btcADRESS_XXX"�#{
  "offerPayload": {
"id": "XXXCENSORED_TRADEID_XXX",
"date": XXXCENSORED_TIMESTAMP_XXX,
"ownerNodeAddress": {
  "hostName": "XXXCENSORED_XXX.onion",
  "port": 9999
},
"direction": "BUY",
"price": 0,
"marketPriceMargin": 0.005,
"useMarketBasedPrice": true,
"amount": 700000,
"minAmount": 300000,
"baseCurrencyCode": "BTC",
"counterCurrencyCode": "CHF",
"arbitratorNodeAddresses": [],
"mediatorNodeAddresses": [
  {
    "hostName": "apbp7ubuyezav4hy.onion",
    "port": 9999
  },
  {
    "hostName": "a56olqlmmpxrn5q34itq5g5tb5d3fg7vxekpbceq7xqvfl3cieocgsyd.onion",
    "port": 9999
  },
  {
    "hostName": "sjlho4zwp3gecspf.onion",
    "port": 9999
  }
],
"paymentMethodId": "SEPA",
"makerPaymentAccountId": "XXXCENSORED_XXX",
"offerFeePaymentTxId": "XXXCENSORED_XXX",
"countryCode": "XXXCENSORED_BANKCC_XXX",
"acceptedCountryCodes": [
  "AT",
  "BE",
  "BG",
  "CH",
  "CY",
  "CZ",
  "DE",
  "DK",
  "EE",
  "ES",
  "FI",
  "FR",
  "GB",
  "GR",
  "HR",
  "HU",
  "IE",
  "IS",
  "IT",
  "LI",
  "LT",
  "LU",
  "LV",
  "MC",
  "MT",
  "NL",
  "NO",
  "PL",
  "PT",
  "RO",
  "SE",
  "SI",
  "SK"
],
"bankId": "XXXCENSORED_XXX",
"versionNr": "1.5.3",
"blockHeightAtOfferCreation": XXXCENSORED_XXX,
"txFee": XXXCENSORED_XXX,
"makerFee": 5000,
"isCurrencyForMakerFeeBtc": true,
"buyerSecurityDeposit": 600000,
"sellerSecurityDeposit": 600000,
"maxTradeLimit": 1000000,
"maxTradePeriod": XXXCENSORED_XXX,
"useAutoClose": false,
"useReOpenAfterAutoClose": false,
"lowerClosePrice": 0,
"upperClosePrice": 0,
"isPrivateOffer": false,
"extraDataMap": {
  "capabilities": "0, 1, 2, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16",
  "accountAgeWitnessHash": "XXXCENSORED_XXX"
},
"protocolVersion": 3
  },
  "tradeAmount": 300000,
  "tradePrice": XXXCENSORED_XXX,
  "takerFeeTxID": "XXXCENSORED_XXX",
  "buyerNodeAddress": {
"hostName": "XXXCENSORED_XXX.onion",
"port": 9999
  },
  "sellerNodeAddress": {
"hostName": "XXXCENSORED_XXX.onion",
"port": 9999
  },
  "mediatorNodeAddress": {
"hostName": "sjlho4zwp3gecspf.onion",
"port": 9999
  },
  "isBuyerMakerAndSellerTaker": true,
  "makerAccountId": "XXXCENSORED_XXX",
  "takerAccountId": "XXXCENSORED_XXX",
  "makerPaymentAccountPayload": {
"holderName": "XXXCENSORED_XXX",
"iban": "XXXCENSORED_XXX",
"bic": "XXXCENSORED_XXX",
"email": "",
"acceptedCountryCodes": [
  "AT",
  "BE",
  "BG",
  "CH",
  "CY",
  "CZ",
  "DE",
  "DK",
  "EE",
  "ES",
  "FI",
  "FR",
  "GB",
  "GR",
  "HR",
  "HU",
  "IE",
  "IS",
  "IT",
  "LI",
  "LT",
  "LU",
  "LV",
  "MC",
  "MT",
  "NL",
  "NO",
  "PL",
  "PT",
  "RO",
  "SE",
  "SI",
  "SK"
],
"countryCode": "XXXCENSORED_XXX",
"paymentMethodId": "SEPA",
"id": "XXXCENSORED_XXX",
"maxTradePeriod": -1
  },
  "takerPaymentAccountPayload": {
"holderName": "XXXCENSORED_XXX",
"iban": "XXXCENSORED_XXX",
"bic": "XXXCENSORED_XXX",
"email": "",
"acceptedCountryCodes": [
  "AT",
  "BE",
  "BG",
  "CH",
  "CY",
  "CZ",
  "DE",
  "DK",
  "EE",
  "ES",
  "FI",
  "FR",
  "GB",
  "GR",
  "HR",
  "HU",
  "IE",
  "IS",
  "IT",
  "LI",
  "LT",
  "LU",
  "LV",
  "MC",
  "MT",
  "NL",
  "NO",
  "PL",
  "PT",
  "RO",
  "SE",
  "SI",
  "SK"
],
"countryCode": "XXXCENSORED_XXX",
"paymentMethodId": "SEPA",
"id": "XXXCENSORED_XXX",
"maxTradePeriod": -1
  },
  "makerPayoutAddressString": "XXXCENSORED_XXX",
  "takerPayoutAddressString": "XXXCENSORED_XXX",
  "lockTime": 669272,
  "refundAgentNodeAddress": {
"hostName": "3z5jnirlccgxzoxc6zwkcgwj66bugvqplzf6z2iyd5oxifiaorhnanqd.onion",
"port": 9999
  }
}*@
```

### PR DESCRIPTION
**Pull request by ghost**
- Removes sensitive payment account information and trader chat messages from closed trades and disputes.
- User can specify in preferences how long to wait before the information is removed, defaults to 20 days.
- The feature is enabled after the user clicks `I UNDERSTAND` shown in the dialog below.
 
Closes #5396

----

### Screenshots:

image

**Settings:**

image

**The highlighted payment account details get removed:**

image

**Trade contract showing payment account information removed.**

image

image

---

### Invocation:

The list of closed trades need to be occasionally evaluated to see if any are eligible for having their sensitive data cleared.  This is primarily done at app initialization, but also is done periodically whenever a trade is added to the list shown in portfolio history.  That is to handle the case of clients which do not restart daily.

Likewise, the list of dispute tickets is evaluated at initialization and also whenever a dispute ticket closed message is processed.

Whenever the evaluation is done, the following log messages are written:

    INFO  b.c.t.ClosedTradableManager: checking closed trades eligibility for having sensitive data cleared 
    INFO  b.c.s.dispute.DisputeManager: ArbitrationManager checking closed disputes eligibility for having sensitive data cleared 
    INFO  b.c.s.dispute.DisputeManager: MediationManager checking closed disputes eligibility for having sensitive data cleared 
    INFO  b.c.s.dispute.DisputeManager: RefundManager checking closed disputes eligibility for having sensitive data cleared

If any actions are taken, they are logged, for example:

    INFO  b.c.trade.model.bisq_v1.Trade: cleared sensitive data from contract;contractAsJson; of trade rfvUYPIx 
    INFO  b.core.support.dispute.Dispute: cleared sensitive data from chat messages; of dispute for trade IOAMZBWH

### CODE REVIEW

**Comment by pazza83** [comment]
Hi @jmacxx thanks for sorting this out.

just checking trades in mediation and arbitration will not count as completed trades?

Assuming mediated trades are completed once payout occurs.

What about arbitrated trades? What point do these become completed after payment is made to donation address?

**Comment by pazza83** [comment]
Also just wondering how this would affect how a duplicate offer creation when a previous offer has had their payload info removed?

**Comment by ghost** [comment]
Answering your questions @pazza83 :

> just checking trades in mediation and arbitration will not count as completed trades?
> Assuming mediated trades are completed once payout occurs.

Correct, only trades that are Closed are considered.

> What about arbitrated trades? What point do these become completed after payment is made to donation address?

Same, they transition to Closed once the arbitration ticket has been closed.

> Also just wondering how this would affect how a duplicate offer creation when a previous offer has had their payload info removed?

No effect as offers do not contain account payloads or contracts, only trades do.  

---

There's some work remaining to sanitize account payloads from closed dispute tickets which looks like the implementation will need refactoring, so moving this back into draft for a couple of days.

**Comment by pazza83** [comment]
> No effect as offers do not contain account payloads or contracts, only trades do.

Just so I am clear, I am assuming the payload is link to the local payment account (trading account) used to fund the trade?

When an offer moves to 'history' currently it still has the information associated to what payment account (trading account) was used to fund the trade associated to it. Will that still be the case once the payload has been cleared?

What I am getting at is once the trade contract details have been cleared, a local user should still not be able just to right click duplicate on a historic trade to see what local payment account (trading account) was used to fund the trade.

**Comment by ghost** [comment]
I think we still want to maintain the ability to duplicate an offer as it is quite frequently used.  

Severing past offers' `makerPaymentAccountId` link to the account used would make duplicating an offer more of a multi-step process similar to offer entry, and is not what this PR is about.  

What is being done here is **removing counterparties personal information from each Bisq instance**.  Yes it also removes copies of your own personal information from trades, which is good too.

**Review by ripcurlx — APPROVED** [review]
utACK - Code changes are looking fine 👍

**Comment by ripcurlx** [comment]
Bildschirmfoto 2022-02-24 um 11 25 20

@jmacxx I think we should add an empty state to the chat popup if all messages have been deleted. Otherwise it looks like an error if someone doesn't remember this privacy feature. WDYT?

**Comment by ripcurlx** [comment]
> Bildschirmfoto 2022-02-24 um 11 25 20
> 
> @jmacxx I think we should add an empty state to the chat popup if all messages have been deleted. Otherwise it looks like an error if someone doesn't remember this privacy feature. WDYT?

The additional weird thing about this is that it was a mediation case from today which seems to have lost all chat messages. It only happened on this side of the dispute. The other side still has the messages.  Also the arbitrator still sees the messages.  Application setting is on the default 20 days.

**Comment by ripcurlx** [comment]
I do see the expected logs e.g. for `30534` 
```
Feb.-09 11:12:16.044 [JavaFX Application Thread] INFO  b.c.s.d.Dispute: cleared sensitive data from contract;contractAsJson;chat messages; of dispute for trade 30534 
```
But not for the recent trade `4900341`, so I'm not sure how the chat messages were deleted.

**Comment by ripcurlx** [comment]
Contract details are still available.

**Comment by ghost** [comment]
> But not for the recent trade `4900341`, so I'm not sure how the chat messages were deleted.

I have not come across that issue before.  I'm trying to reproduce it, so far not been able to.

I wonder if anything can be seen in your log when it removed the chat messages from `4900341`?  I would expect a log message about it.

File: core/src/main/java/bisq/core/support/dispute/DisputeManager.java [hunk 1]
```diff
@@ -298,6 +302,15 @@ public boolean isTrader(Dispute dispute) {
         return disputeList.stream().filter(e -> e.getTradeId().equals(tradeId)).findAny();
     }
 
+    public void maybeClearSensitiveData() {
+        log.info("{} checking closed disputes eligibility for having sensitive data cleared", super.getClass().getSimpleName());
+        Instant safeDate = closedTradableManager.getSafeDateForSensitiveDataClearing();
+        getDisputeList().getList().stream()
+                .filter(e -> e.isClosed())
+                .filter(e -> e.getTradeDate().toInstant().isBefore(safeDate))
```
Comment (ripcurlx) [inline core/src/main/java/bisq/core/support/dispute/DisputeManager.java:310]: Maybe you want to use: 
```
.filter(Dispute::isClosed)
.filter(e -> e.getOpeningDate().toInstant().isBefore(safeDate))
```
instead?

**Comment by ripcurlx** [comment]
> > But not for the recent trade `4900341`, so I'm not sure how the chat messages were deleted.
> 
> I have not come across that issue before. I'm trying to reproduce it, so far not been able to.
> 
> I wonder if anything can be seen in your log when it removed the chat messages from `4900341`? I would expect a log message about it.

ah - I think I know what happened. I might have missed the log, but the trade I've put into mediation was already quite old. So at the moment when the dispute was closed, all chat messages were cleared as you are using the trade date instead of the dispute opening date. See https://github.com/bisq-network/bisq/pull/6001/files#r814853854

**Comment by pazza83** [comment]
Tested this and it works well. Many thanks.

**Pull request merged**